### ISSUE TITLE
ReLU and Softmax Activation Functions

### ISSUE THREAD

**Issue by Kulbear** [Kulbear/Schedule#1]
## Refresher: The Sigmoid Function

The sigmoid function has been widely used in machine learning intro materials, especially for the logistic regression and some basic neural network implementations. However, you may need to know that the sigmoid function is not your only choice for the activation function and it does have drawbacks.

![Some sigmoid functions compared.](https://cloud.githubusercontent.com/assets/14886380/22743102/1ddd6a88-ee54-11e6-98ea-6b67011e091b.png)

Let's see what a sigmoid function could benefit us.

![The most common sigmoid function.](https://cloud.githubusercontent.com/assets/14886380/22743129/36eee4b6-ee54-11e6-8163-35c7c641d226.png)

By taking this formula, we can get the derivative of the sigmoid function, note that for shortening the formula, here f(x) is the sigmoid function.

![The derivative of the sigmoid function](https://cloud.githubusercontent.com/assets/14886380/22743164/58665ac0-ee54-11e6-91fb-8ec818752480.png)

Despite that, such a result from the derivate is easy to calculate and save times for building models, the sigmoid function actually forces your model "losing" knowledge from the data. Why? Think about the possible maximum value of the derivative of a sigmoid function.

![Visuals for the sigmoid function and its derivative](https://cloud.githubusercontent.com/assets/14886380/22743184/6c2cef4c-ee54-11e6-93f9-9febd1056ce1.png)

For the backpropagation process in a neural network, it means that your errors will be squeezed by (at least) a quarter at each layer. Therefore, deeper your network is, more knowledge from the data will be "lost". Some “big” errors we get from the output layer might not be able to affect the synapses weight of a neuron in a relatively shallow layer much ("shallow" means it's close to the input layer).

Due to this, sigmoids have fallen out of favor as activations on hidden units.

Addition notes from Wikipedia (just FYI, in case you are interested in):

> Besides the [logistic function](https://en.wikipedia.org/wiki/Logistic_function), sigmoid functions include the ordinary [arctangent](https://en.wikipedia.org/wiki/Inverse_trigonometric_function), the [hyperbolic tangent](https://en.wikipedia.org/wiki/Hyperbolic_function), the [Gudermannian function](https://en.wikipedia.org/wiki/Gudermannian_function), and the [error function](https://en.wikipedia.org/wiki/Error_function), but also the [generalised logistic function](https://en.wikipedia.org/wiki/Generalised_logistic_function) and [algebraic functions](https://en.wikipedia.org/wiki/Algebraic_function)

----

## Rectified Linear Units

Instead of sigmoids, most recent deep learning networks use rectified linear units (ReLUs) for the hidden layers. A rectified linear unit has output 0 if the input is less than 0, and raw output otherwise. That is, if the input is greater than 0, the output is equal to the input. ReLUs' machinery is more like a real neuron in your body.

![The ReLU function](https://cloud.githubusercontent.com/assets/14886380/22743194/73ca0834-ee54-11e6-903f-a7efd247406b.png)

ReLU activations are the simplest non-linear activation function you can use, obviously. When you get the input is positive, the derivative is just 1, so there isn't the squeezing effect you meet on backpropagated errors from the sigmoid function. [Research has shown](http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf) that ReLUs result in much faster training for large networks. Most frameworks like TensorFlow and TFLearn make it simple to use ReLUs on the the hidden layers, so you won't need to implement them yourself.

However, such a simple solution is not always a perfect solution. From [Andrej Karpathy's CS231n course](http://cs231n.github.io/neural-networks-1/#nn):

> Unfortunately, ReLU units can be fragile during training and can “die”. For example, a large gradient flowing through a ReLU neuron could cause the weights to update in such a way that the neuron will never activate on any datapoint again. If this happens, then the gradient flowing through the unit will forever be zero from that point on. That is, the ReLU units can irreversibly die during training since they can get knocked off the data manifold. For example, you may find that as much as 40% of your network can be “dead” (i.e. neurons that never activate across the entire training dataset) if the learning rate is set too high. With a proper setting of the learning rate this is less frequently an issue.

----

## Softmax

The sigmoid function can be applied easily, the ReLUs will not vanish the effect during your training process. However, when you want to deal with classification problems, they cannot help much. Simply speaking, the sigmoid function can only handle two classes, which is not what we expect.

![How does a softmax function work](https://cloud.githubusercontent.com/assets/14886380/22743228/91be8964-ee54-11e6-9a59-c55ea02a9146.png)

The softmax function squashes the outputs of each unit to be between 0 and 1, just like a sigmoid function. But it also divides each output such that the total sum of the outputs is equal to 1 (check it on the figure above). 

The output of the softmax function is equivalent to a categorical probability distribution, it tells you the probability that any of the classes are true.

Mathematically the softmax function is shown below, where z is a vector of the inputs to the output layer (if you have 10 output units, then there are 10 elements in z). And again, j indexes the output units, so j = 1, 2, ..., K.

![The softmax function](https://cloud.githubusercontent.com/assets/14886380/22743247/9eb7c856-ee54-11e6-98ca-a7e03120b1f8.png)

The material below is taken from a Udacity course.

To understand this better, think about training a network to recognize and classify handwritten digits from images. The network would have ten output units, one for each digit 0 to 9. Then if you fed it an image of a number 4 (see below), the output unit corresponding to the digit 4 would be activated.

![Image from  [the MNIST dataset](http://yann.lecun.com/exdb/mnist/)
](https://cloud.githubusercontent.com/assets/14886380/22743262/ab034450-ee54-11e6-9a5f-e3e3e39ef9cd.png)

Building a network like this requires 10 output units, one for each digit. Each training image is labeled with the true digit and the goal of the network is to predict the correct label. So, if the input is an image of the digit 4, the output unit corresponding to 4 would be activated, and so on for the rest of the units.

For the example image above, the output of the softmax function might look like:

![Example softmax output for a network predicting the digit shown above](https://cloud.githubusercontent.com/assets/14886380/22743284/b79be10e-ee54-11e6-94fb-e9832358e0c1.png)

The image looks the most like the digit 4, so you get a lot of probability there. However, this digit also looks somewhat like a 7 and a little bit like a 9 without the loop completed. So, you get the most probability that it's a 4, but also some probability that it's a 7 or a 9.

The softmax can be used for any number of classes. It's also used for hundreds and thousands of classes, for example in object recognition problems where there are hundreds of different possible objects.

## Reference
- All formulas are generated by [HostMath](http://www.hostmath.com/)
- [Sigmoid function - Wikipedia](https://en.wikipedia.org/wiki/Sigmoid_function)
- [ImageNet Classification with Deep Convolutional
Neural Networks](http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf)
- [The MNIST database by Yann LeCun](http://yann.lecun.com/exdb/mnist/)
